### PR TITLE
Reduce number of API operations

### PR DESCRIPTION
**Pull request by mghildiy**

### CODE REVIEW

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 1]
```diff
@@ -58,6 +59,17 @@ public Object clone() throws CloneNotSupportedException {
       Map<String,List<String>> awsDimensionSelectRegex;
       String help;
       boolean cloudwatchTimestamp;
+
+      List<String> allStats() {
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:63]: This name is misleading. Per the name I'd expect it to be the defaults if nothing is user specified, not copying in the configuration.

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 2]
```diff
@@ -377,120 +376,157 @@ private String help(MetricRule rule, String unit, String statistic) {
...
-        if (rule.awsNamespace.equals("AWS/DynamoDB")
-                && rule.awsDimensions.contains("GlobalSecondaryIndexName")
-                && brokenDynamoMetrics.contains(rule.awsMetricName)) {
-            baseName += "_index";
+    private static String queryId(MetricRule rule, List<Dimension> dimensions, String stat) {
+      String idComponents = rule.awsNamespace +
+              rule.awsMetricName +
+              dimensions.hashCode() +
+              stat;
+      // We cant't use the concatenated components as the ID directly because AWS will not accept special chars as ID
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:384]: What happens if there's a collision? It'd be safer to use a counter.

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 3]
```diff
@@ -377,120 +376,157 @@ private String help(MetricRule rule, String unit, String statistic) {
...
+      // Also, the ID must start with a letter, not a digit
+      return "q" + Integer.toUnsignedString(idComponents.hashCode());
+    }
+
+    private static MetricDataQuery metricDataQuery(MetricRule rule, List<Dimension> dimensions, String stat) {
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:389]: Simple functions like this that are only used once make it hard to piece the code together, it'd be better to inline it. This is part of a chain of at least 5 such functions, requiring me to bounce around the file to try and understand the code.

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 4]
```diff
@@ -377,120 +376,157 @@ private String help(MetricRule rule, String unit, String statistic) {
...
+
+      return new MetricDataQuery()
+              .withId(queryId(rule, dimensions, stat))
+              .withMetricStat(metricStat);
+    }
+
+    private static List<MetricDataQuery> queriesForRuleDimensions(MetricRule rule, List<Dimension> dimensions) {
+      List<MetricDataQuery> queries = new ArrayList<>();
+
+      // Cost of fetching data from AWS stays the same
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:408]: What is this comment for?

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 5]
```diff
@@ -377,120 +376,157 @@ private String help(MetricRule rule, String unit, String statistic) {
...
+
+      return requests;
+    }
+
+  private Map<String, MetricDataResult> fetchResults(ActiveConfig config, long start) {
+    Map<String, MetricDataResult> idsToResults = new HashMap<>();
+    for (MetricRule rule : config.rules) {
+      final List<GetMetricDataRequest> getMetricDataRequests = requestForRule(rule, config, start);
+      for (GetMetricDataRequest request : getMetricDataRequests) {
+        if (request == null) {
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:452]: How can this be null?

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 6]
```diff
@@ -377,120 +376,157 @@ private String help(MetricRule rule, String unit, String statistic) {
...
-        }
-        if (!averageSamples.isEmpty()) {
-          mfs.add(new MetricFamilySamples(baseName + "_average", Type.GAUGE, help(rule, unit, "Average"), averageSamples));
-        }
-        for (Map.Entry<String, ArrayList<MetricFamilySamples.Sample>> entry : extendedSamples.entrySet()) {
-          mfs.add(new MetricFamilySamples(baseName + "_" + safeName(toSnakeCase(entry.getKey())), Type.GAUGE, help(rule, unit, entry.getKey()), entry.getValue()));
-        }
+      for (String statType : statsTypeToSamples.keySet()) {
+        mfs.add(new MetricFamilySamples(baseName + "_" + safeName(statType).toLowerCase(),
+                Type.GAUGE, help(rule, "UNKNOWN", "Sum"), statsTypeToSamples.get(statType))
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:525]: A unit that is always unknown isn't very useful. Also Sum isn't right.

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 7]
```diff
@@ -377,120 +376,157 @@ private String help(MetricRule rule, String unit, String statistic) {
...
+      String jobName = safeName(rule.awsNamespace.toLowerCase());
+      String baseName = safeName(rule.awsNamespace.toLowerCase() + "_" + toSnakeCase(rule.awsMetricName));
+
+      if (rule.awsNamespace.equals("AWS/DynamoDB")
+              && rule.awsDimensions.contains("GlobalSecondaryIndexName")
+              && brokenDynamoMetrics.contains(rule.awsMetricName)) {
+        baseName += "_index";
+      }
+
+      for (List<Dimension> dimensions : getDimensions(rule, config.client)) {
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:490]: getDimensions makes API calls, which were already made above. We shouldn't be calling it twice.

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 8]
```diff
@@ -377,120 +376,157 @@ private String help(MetricRule rule, String unit, String statistic) {
...
-          }
 
-          Long timestamp = null;
-          if (rule.cloudwatchTimestamp) {
-            timestamp = dp.getTimestamp().getTime();
-          }
+          MetricFamilySamples.Sample sample = new MetricFamilySamples.Sample(
+                  baseName + "_" + safeName(stat).toLowerCase(),
+                  labelNames, labelValues,
+                  result.getValues().get(0), result.getTimestamps().get(0).getTime()
```
Comment (brian-brazil) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:512]: Timestamps should only be used it configured.

**Review by brian-brazil — COMMENTED** [review]
This needs to be brought up to date with the current codebase, it's missing resource info for example.

**Comment by mghildiy** [comment]
It seems I would need to re-look at the entire code. Most of these changes are legacy changes. It would be better if I take from latest code and try to make Johannes' changes o top of that with above review comments considered.

**Comment by brian-brazil** [comment]
Yeah, might be best to take it as a guide for what needs changing.

**Comment by matthiasr** [comment]
@mghildiy is there anything I can do to help you with this?

**Comment by matthiasr** [comment]
(gave this a more descriptive title)

**Comment by matthiasr** [comment]
I'm going to close this for now, since it is not moving forward. Please reopen (or open a new one) if you want to work on this!

**Pull request closed without merging**